promunifi/site.go

```diff
@@ -0,0 +1,155 @@
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"golift.io/unifi"
+)
+
+// Each site has five subsystems.
+// Some of them share data, but generally each one has its own data set.
+const (
+	subsystemLAN  = "lan"
+	subsystemVPN  = "vpn"
+	subsystemWWW  = "www"
+	subsystemWLAN = "wlan"
+	subsystemWAN  = "wan"
+)
+
+type site struct {
+	NumUser               *prometheus.Desc
+	NumGuest              *prometheus.Desc
+	NumIot                *prometheus.Desc
+	TxBytesR              *prometheus.Desc
+	RxBytesR              *prometheus.Desc
+	NumAp                 *prometheus.Desc
+	NumAdopted            *prometheus.Desc
+	NumDisabled           *prometheus.Desc
+	NumDisconnected       *prometheus.Desc
+	NumPending            *prometheus.Desc
+	NumGw                 *prometheus.Desc
+	NumSw                 *prometheus.Desc
+	NumSta                *prometheus.Desc
+	Latency               *prometheus.Desc
+	Drops                 *prometheus.Desc
+	Uptime                *prometheus.Desc
+	XputUp                *prometheus.Desc
+	XputDown              *prometheus.Desc
+	SpeedtestPing         *prometheus.Desc
+	RemoteUserNumActive   *prometheus.Desc
+	RemoteUserNumInactive *prometheus.Desc
+	RemoteUserRxBytes     *prometheus.Desc
+	RemoteUserTxBytes     *prometheus.Desc
+	RemoteUserRxPackets   *prometheus.Desc
+	RemoteUserTxPackets   *prometheus.Desc
+}
+
+func descSite(ns string) *site {
+	if ns += "_site_"; ns == "_site_" {
+		ns = "site_"
+	}
+	labels := []string{"subsystem", "status", "name", "desc", "site_name"}
+
+	return &site{
+		NumUser:               prometheus.NewDesc(ns+"num_user_total", "Number of Users", labels, nil),
+		NumGuest:              prometheus.NewDesc(ns+"num_guest_total", "Number of Guests", labels, nil),
+		NumIot:                prometheus.NewDesc(ns+"num_iot_total", "Number of IoT Devices", labels, nil),
+		TxBytesR:              prometheus.NewDesc(ns+"transmit_rate_bytes", "Bytes Transmit Rate", labels, nil),
+		RxBytesR:              prometheus.NewDesc(ns+"receive_rate_bytes", "Bytes Receive Rate", labels, nil),
+		NumAp:                 prometheus.NewDesc(ns+"num_ap_total", "Access Point Count", labels, nil),
+		NumAdopted:            prometheus.NewDesc(ns+"num_adopted_total", "Adoption Count", labels, nil),
+		NumDisabled:           prometheus.NewDesc(ns+"num_disabled_total", "Disabled Count", labels, nil),
+		NumDisconnected:       prometheus.NewDesc(ns+"num_disconnected_total", "Disconnected Count", labels, nil),
+		NumPending:            prometheus.NewDesc(ns+"num_pending_total", "Pending Count", labels, nil),
+		NumGw:                 prometheus.NewDesc(ns+"num_gateways_total", "Gateway Count", labels, nil),
+		NumSw:                 prometheus.NewDesc(ns+"num_switches_total", "Switch Count", labels, nil),
+		NumSta:                prometheus.NewDesc(ns+"num_stations_total", "Station Count", labels, nil),
+		Latency:               prometheus.NewDesc(ns+"latency_ms", "Latency", labels, nil),
```

Comment:
We typically convert milliseconds to seconds when exposing data in Prometheus. This makes it easier to compare data cross-metric. Then we display the unit (us, ms, s) automatically in things like Grafana.